### PR TITLE
Add support for Launch Templates, custom AMI and more features for Managed Nodegroups

### LINKED ISSUE #2551: Custom AMI and Launch Template support for Managed Nodegroups
EKS now supports specifying a custom AMI and providing a launch template for managed nodegroups. This will allow eksctl to support features that weren't possible before for managed nodegroups.

Add support for missing features using the launch template capability, including the following:
- Custom AMI (`ami`)
- `instancePrefix` and `instanceName` to configure the EC2 instance name for a nodegroup
- `securityGroups.attachIDs` to add custom security groups
- `ebsOptimized`, `volumeType`, `volumeName`, `volumeEncrypted`, `volumeKmsKeyID` and `volumeIOPS`
- `maxPodsPerNode`
- `preBootstrapCommands` and `overrideBootstrapCommand`
- `disableIMDSv1`

Also add support for providing a custom launch template to create a managed nodegroup.

### PR DESCRIPTION
**Pull request by cPu1**
### Description
Adds launch template support for managed nodegroups. Also adds support for the following features for managed nodegroups:

- Custom AMI (`ami`)
- `instancePrefix` and `instanceName` to configure the EC2 instance name for a nodegroup
- `securityGroups.attachIDs` to add custom security groups
- `ebsOptimized`, `volumeType`, `volumeName`, `volumeEncrypted`, `volumeKmsKeyID` and `volumeIOPS`
- `maxPodsPerNode`
- `preBootstrapCommands` and `overrideBootstrapCommand`
- `disableIMDSv1`

To use the launch templates features, specify an existing launch template:

```yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig
metadata:
  name: cluster
  region: us-west-2
managedNodeGroups:
- name: launch-template-ng
  launchTemplate:
    id: lt-<id>
    version: "1" # (optional) defaults to the default launch template version
```

TODO:
- [x] Add documentation

Closes #2551

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Added labels for change area (e.g. `area/nodegroup`), target version (e.g. `version/0.12.0`) and kind (e.g. `kind/improvement`)
- [x] Make sure the title of the PR is a good description that can go into the release notes

### CODE REVIEW

File: pkg/cfn/builder/managed_launch_template_test.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		err := stack.AddAllResources()
+		Expect(err).ToNot(HaveOccurred())
+		bytes, err := stack.RenderJSON()
+		Expect(err).ToNot(HaveOccurred())
+
+		template, err := goformation.ParseJSON(bytes)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(template).ToNot(BeNil())
+
+		actual, err := json.Marshal(template.Resources)
```
Comment (michaelbeaumont) [inline pkg/cfn/builder/managed_launch_template_test.go:51]: Am I reading this correctly that the tests unmarshal and marshal and then compare the json? Why not match on the structs?

File: pkg/cfn/builder/managed_nodegroup_test.go [hunk 2]
```diff
@@ -71,27 +71,27 @@ func TestManagedPolicyResources(t *testing.T) {
...
 
 			ng := api.NewManagedNodeGroup()
 			ng.IAM.WithAddonPolicies = tt.addons
 			ng.IAM.AttachPolicyARNs = prefixPolicies(tt.attachPolicyARNs...)
 
-			stack := NewManagedNodeGroup(clusterConfig, ng, "iam-test")
+			stack := NewManagedNodeGroup(clusterConfig, ng, nil, "iam-test")
 			err := stack.AddAllResources()
-			assert.Nil(err)
+			require.Nil(err)
```
Comment (michaelbeaumont) [inline pkg/cfn/builder/managed_nodegroup_test.go:83]: :hushed:  What's the difference?

Comment (cPu1) [reply pkg/cfn/builder/managed_nodegroup_test.go:83]: The methods in `require` stop the test (calling `FailNow`) when an assertion fails.

Comment (michaelbeaumont) [reply pkg/cfn/builder/managed_nodegroup_test.go:83]: ah ok so probably it should always be `require` :+1:

File: pkg/cfn/builder/managed_launch_template_test.go (hunk 1 above)
Comment (cPu1) [reply pkg/cfn/builder/managed_launch_template_test.go:51]: That's right. The unmarshalling (`goformation.ParseJSON(bytes)`) and marshalling is actually testing goformation's ability to preserve isomorphism, to ensure no types or values are lost or are interpreted differently. 

>Why not match on the structs?

Matching on the structs will require more code than matching on the expected JSON output. This is also close to an end-to-end test.

Comment (michaelbeaumont) [reply pkg/cfn/builder/managed_launch_template_test.go:51]: makes sense!

**Comment by michaelbeaumont** [comment]
Can we also add some integration tests?

**Comment by cPu1** [comment]
> Can we also add some integration tests?

I was planning to but couldn't get to it. I'll address it in a separate PR. WDYT?

File: userdocs/src/usage/eks-managed-nodes.md [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 An EKS managed node group is an autoscaling group and associated EC2 instances that are managed by AWS for an Amazon EKS cluster. Each node group uses the Amazon EKS-optimized Amazon Linux 2 AMI. Amazon EKS makes it easy to apply bug fixes and security patches to nodes, as well as update them to the latest Kubernetes versions. Each node group launches an autoscaling group for your cluster, which can span multiple AWS VPC availability zones and subnets for high-availability.
 
+**NEW** [Launch Template support for managed nodegroups](launch-template-support.md)
```
Comment (michaelbeaumont) [inline userdocs/src/usage/eks-managed-nodes.md:7]: Did you forget to commit this?

Comment (cPu1) [reply userdocs/src/usage/eks-managed-nodes.md:7]: I did. Thanks!

**Comment by kkl-acies** [comment]
Did this PR remove the network interface from the created launch template? If yes why?

**Pull request merged**